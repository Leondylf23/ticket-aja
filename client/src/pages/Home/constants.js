export const GET_TICKETS_DATA = 'Home/GET_TICKETS_DATA';
export const GET_MY_TICKETS_DATA = 'Home/GET_MY_TICKETS_DATA';
export const SET_TICKETS_DATA = 'Home/SET_TICKETS_DATA';