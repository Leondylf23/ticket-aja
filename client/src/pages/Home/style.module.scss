@import '@styles/base';

@include mainContainer {
  .title {
    margin-bottom: 0.6rem;
  }
  .createBtnContainer {
    @include directionX;

    justify-content: end;
    margin-bottom: 0.6rem;

    .button {
      @include button;
    }
  }
  .searchContainer {
    @include directionY;
    @include padding(0.6rem);

    background-color: var(--color-bg);
    width: 100%;
    margin-bottom: 1rem;
    border-radius: 10px;
    box-shadow: 5px 5px 15px rgba(0, 0, 0, 0.2);

    .title {
      font-size: 1rem;
    }
    .input {
      @include input;
      width: 100%;
    }
  }
  .dataContainer {
    @include directionX;

    flex-grow: 1;

    .dataGrid {
      width: 100%;

      .innerDataGrid {
        @include grid(1, 0.6rem);
        width: 100%;

        @include apply-at-sm-size {
          @include grid(2, 0.6rem);
        }
        @include apply-at-lg-size {
          @include grid(3, 0.6rem);
        }
      }
    }
    .empty {
      @include containerCenter;

      width: 100%;
      flex-grow: 1;
    }
  }
}
