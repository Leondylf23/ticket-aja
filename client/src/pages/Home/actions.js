import { GET_MY_TICKETS_DATA, GET_TICKETS_DATA, SET_TICKETS_DATA } from "./constants";

export const getTicketsData = (formData) => ({
    type: GET_TICKETS_DATA,
    formData
});

export const getMyTicketsData = (formData) => ({
    type: GET_MY_TICKETS_DATA,
    formData
});

export const setTicketsData = (data) => ({
    type: SET_TICKETS_DATA,
    data
});