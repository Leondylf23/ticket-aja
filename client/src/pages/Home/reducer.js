import { produce } from 'immer';
import { SET_TICKETS_DATA } from './constants';

export const initialState = {
    tickets: []
};

export const storedKey = [];

const homeReducer = (state = initialState, action) =>
    produce(state, (draft) => {
        switch (action.type) {
            case SET_TICKETS_DATA:
                draft.tickets = action.data;
                break;
        }
    });

export default homeReducer;
