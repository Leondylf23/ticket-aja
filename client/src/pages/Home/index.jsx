import { useEffect, useState } from 'react';
import { useDispatch, connect } from 'react-redux';
import { FormattedMessage, useIntl } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { useNavigate } from 'react-router-dom';
import PropTypes from 'prop-types';

import TicketCard from '@components/TicketCard';
import { selectTicketData } from './selectors';
import { getUserDataDecrypt } from '@utils/allUtils';
import { getMyTicketsData, getTicketsData } from './actions';

import clasess from './style.module.scss';
import { selectUserData } from '@containers/Client/selectors';

const Home = ({ tickets, userData }) => {
  const dispatch = useDispatch();
  const intl = useIntl();
  const navigate = useNavigate();

  const [data, setData] = useState([]);
  const [search, setSearch] = useState('');
  const [isBusiness, setIsBusiness] = useState(false);
  const [timerId, setTimerId] = useState(null);

  const getTicketFromApi = () => {
    const user = getUserDataDecrypt(userData);

    if (user && user?.role === 'business') {
      setIsBusiness(true);
      dispatch(getMyTicketsData({...(search !== '' && { ticketName: search })}));
    } else {
      setIsBusiness(false);
      dispatch(getTicketsData({...(search !== '' && { ticketName: search })}));
    }
  };

  useEffect(() => {
    if (tickets) {
      setData(tickets);
    }
  }, [tickets]);
  useEffect(() => {
    getTicketFromApi();
  }, [userData]);
  useEffect(() => {
    if (timerId) {
      clearTimeout(timerId);
    }

    setTimerId(setTimeout(() => {
      getTicketFromApi();
      setTimerId(null);
    }, 500));
  }, [search]);

  return (
    <div className={clasess.mainContainer}>
      <h1 className={clasess.title}><FormattedMessage id='home_title' /></h1>
      {isBusiness && <div className={clasess.createBtnContainer}>
        <button className={clasess.button} onClick={() => navigate('/ticketcreation')}><FormattedMessage id='home_create_new' /></button>
      </div>}
      <div className={clasess.searchContainer}>
        <h4 className={clasess.title}><FormattedMessage id='home_title' /></h4>
        <input type="text" className={clasess.input} placeholder={intl.formatMessage({ id: 'home_search_placeholder' })} onChange={(e) => setSearch(e.target.value)} />
      </div>
      <div className={clasess.dataContainer}>
        {data.length > 0 ? <div className={clasess.dataGrid}>
          <div className={clasess.innerDataGrid}>
            {data.map(ticket =>
              <TicketCard data={ticket} isBusiness={isBusiness} key={ticket?.id} />
            )}
          </div>
        </div> : <div className={clasess.empty}>
          <h3><FormattedMessage id='empty_data' /></h3>
        </div>}
      </div>
    </div>
  );
};

Home.propTypes = {
  userData: PropTypes.string,
  tickets: PropTypes.array
}

const mapStateToProps = createStructuredSelector({
  userData: selectUserData,
  tickets: selectTicketData
});

export default connect(mapStateToProps)(Home);
