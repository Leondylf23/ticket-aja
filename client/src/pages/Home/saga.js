import { takeLatest, call, put } from 'redux-saga/effects';

import { showPopup, setLoading } from '@containers/App/actions';
import { GET_MY_TICKETS_DATA, GET_TICKETS_DATA } from './constants';
import { getAllTickets, getMyAllTickets } from '@domain/api';
import { setTicketsData } from './actions';

function* doGetTicketData({ formData }) {
    yield put(setLoading(true));

    try {
        const res = yield call(getAllTickets, formData);
        yield put(setTicketsData(res?.data));
    } catch (error) {
        yield put(showPopup());
    }

    yield put(setLoading(false));
}

function* doGetMyTicketData({ formData }) {
    yield put(setLoading(true));
    
    try {
        const res = yield call(getMyAllTickets, formData);
        yield put(setTicketsData(res?.data));
    } catch (error) {
        yield put(showPopup());
    }

    yield put(setLoading(false));
}

export default function* homeSaga() {
    yield takeLatest(GET_TICKETS_DATA, doGetTicketData);
    yield takeLatest(GET_MY_TICKETS_DATA, doGetMyTicketData);
}
