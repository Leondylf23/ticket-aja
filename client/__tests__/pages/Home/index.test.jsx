import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import RouterDom from 'react-router-dom';

import Home from '@pages/Home';
import store from '@store';
import Language from '@containers/Language';


jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux')
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn()
}));

jest.mock('reselect', () => ({
    ...jest.requireActual('reselect')
}));

const ParentComponent = (children) => (
    <Provider store={store}>
        <Language>
            {children}
        </Language>
    </Provider>
);

// const navigate = jest.fn().mockName('navigate');

describe('Home Page', () => {
    // beforeEach(() => {
    //     jest.spyOn(RouterDom, 'useNavigate').mockImplementation(() => navigate);
    // });

    test('Rendered', () => {
        const { getByTestId } = render(ParentComponent(
            <Home />
        ));
        const homePage = getByTestId('home-page');
        expect(homePage).toBeInTheDocument();
    });

    test('Should match with snapshot', () => {
        const homePage = render(ParentComponent(
            <Home />
        ));
        expect(homePage).toMatchSnapshot();
      });
});
