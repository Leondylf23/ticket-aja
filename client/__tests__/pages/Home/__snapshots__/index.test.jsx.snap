// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mainContainer"
        data-testid="home-page"
      >
        <h1
          class="title"
        >
          Cari Tiket
        </h1>
        <div
          class="searchContainer"
        >
          <h4
            class="title"
          >
            Cari Tiket
          </h4>
          <input
            class="input"
            placeholder="Cari tiket..."
            type="text"
          />
        </div>
        <div
          class="dataContainer"
        >
          <div
            class="empty"
          >
            <h3>
              Kosong
            </h3>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mainContainer"
      data-testid="home-page"
    >
      <h1
        class="title"
      >
        Cari Tiket
      </h1>
      <div
        class="searchContainer"
      >
        <h4
          class="title"
        >
          Cari Tiket
        </h4>
        <input
          class="input"
          placeholder="Cari tiket..."
          type="text"
        />
      </div>
      <div
        class="dataContainer"
      >
        <div
          class="empty"
        >
          <h3>
            Kosong
          </h3>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
